pkg/roachprod/failureinjection/failures/disk_stall.go

```diff
@@ -0,0 +1,474 @@
+		return err
+	}
+
+	if err = s.Run(ctx, l, s.c.Nodes, `sudo mount /dev/mapper/data1 /mnt/data1`); err != nil {
+		return err
+	}
+
+	if diskStallArgs.RestartNodes {
+		if err = s.StartCluster(ctx, l); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *DmsetupDiskStaller) Inject(ctx context.Context, l *logger.Logger, args FailureArgs) error {
+	nodes := args.(DiskStallArgs).Nodes
+	l.Printf("stalling disk I/O on nodes %d", nodes)
+	return s.Run(ctx, l, nodes, `sudo dmsetup suspend --noflush --nolockfs data1`)
+}
+
+func (s *DmsetupDiskStaller) Recover(
+	ctx context.Context, l *logger.Logger, args FailureArgs,
+) error {
+	diskStallArgs := args.(DiskStallArgs)
+	nodes := diskStallArgs.Nodes
+	l.Printf("unstalling disk I/O on nodes %d", nodes)
+	if err := s.Run(ctx, l, nodes, `sudo dmsetup resume data1`); err != nil {
+		return err
+	}
+	// If the disk stall was injected for long enough that the cockroach process
+	// detected it and shut down the node, then restart it.
+	return forEachNode(nodes, func(n install.Nodes) error {
+		if err := s.PingNode(ctx, l, n); err != nil && diskStallArgs.RestartNodes {
+			l.Printf("failed to connect to n%d, assuming node exited and restarting: %v", n, err)
+			return s.StartNodes(ctx, l, n)
+		}
+		return nil
+	})
+}
+
+func (s *DmsetupDiskStaller) Cleanup(
+	ctx context.Context, l *logger.Logger, args FailureArgs,
+) error {
+	diskStallArgs := args.(DiskStallArgs)
+	if diskStallArgs.RestartNodes {
+		stopOpts := roachprod.DefaultStopOpts()
+		if err := s.StopCluster(ctx, l, stopOpts); err != nil {
+			return err
+		}
+	}
+
+	dev, err := s.DiskDeviceName(ctx, l)
+	if err != nil {
+		return err
+	}
+
+	if err := s.Run(ctx, l, s.c.Nodes, `sudo dmsetup resume data1`); err != nil {
+		return err
+	}
+	if err := s.Run(ctx, l, s.c.Nodes, `sudo umount /mnt/data1`); err != nil {
+		return err
+	}
+	if err := s.Run(ctx, l, s.c.Nodes, `sudo dmsetup remove data1`); err != nil {
+		return err
+	}
+	if err := s.Run(ctx, l, s.c.Nodes, `sudo tune2fs -O has_journal `+dev); err != nil {
+		return err
+	}
+	if err := s.Run(ctx, l, s.c.Nodes, `sudo mount /mnt/data1`); err != nil {
+		return err
+	}
+	// Reinstall snapd.
+	if err := s.Run(ctx, l, s.c.Nodes, `sudo apt-get install -y snapd`); err != nil {
+		return err
+	}
+
+	// When we unmounted the disk in setup, the cgroups controllers may have been removed, re-add them.
+	if err := s.Run(ctx, l, s.c.Nodes, "sudo", "/bin/bash", "-c",
+		`'echo "+cpuset +cpu +io +memory +pids" > /sys/fs/cgroup/cgroup.subtree_control'`); err != nil {
+		return err
+	}
+	if err := s.Run(ctx, l, s.c.Nodes, "sudo", "/bin/bash", "-c",
+		`'echo "+cpuset +cpu +io +memory +pids" > /sys/fs/cgroup/system.slice/cgroup.subtree_control'`); err != nil {
+		return err
+	}
+
+	if diskStallArgs.RestartNodes {
+		if err := s.StartCluster(ctx, l); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *DmsetupDiskStaller) WaitForFailureToPropagate(
```

Comment:
Both of these helpers seem generic; could be moved to `GenericFailure`.

Reply:
Right now they are generic but in the future I plan on adding checks for e.g. replication, range rebalancing, etc.